### ISSUE TITLE
New jobs created with Jenkins REST API are not immediately registered with generic webook

### ISSUE THREAD

**Issue by nuxi** [jenkinsci/generic-webhook-trigger-plugin#312]
This seems alot like #311 but in this case I'm using the REST API and it appears to have worked fine in an older version of the plugin. I have two (almost identical) Jenkins servers, the test server had plugin v2.2.1 and the production server has v2.0.0

On the server with v2.0.0 I was able to create jobs with the REST API and they immediately showed up in the response given by the /generic-webhook-trigger/invoke endpoint. On the server with v2.2.1 they do not show up until I open them and re-save the config.

The script creating the jobs is the same in both cases:
```
    if ! curl -s -f "$JENKINS_URL/job/$NEW/config.xml" -u "$JENKINS_API_KEY" -o /dev/null 2> /dev/null ; then
        # create an empty job first to work around bug JENKINS-64356
        echo "Creating empty job $NEW"
        curl -s -S -f -X POST "$JENKINS_URL/createItem?name=$NEW&mode=hudson.model.FreeStyleProject" \
             -u "$JENKINS_API_KEY" \
             -H "Content-Type: application/x-www-form-urlencoded"
    fi
    
    # now post the new configuration into it
    echo "Posting config to $NEW"
    curl -s -S -f -X POST "$JENKINS_URL/job/$NEW/config.xml" \
         -u "$JENKINS_API_KEY" \
         -H "Content-Type: text/xml" \
         --data-binary @config.xml
```
Note that in order to workaround the workaround for an old jenkins bug, the script technically creates an empty job first and then updates the job to contain the right configuration.

I downgraded the plugin on the server with v2.2.1 to v2.0.0 and it seems to work perfectly.

### Jenkins and plugins versions report

This is from the server that originally had v2.2.1 (although I only captured the data after downgrading it to v2.0.0)
```
Jenkins: 2.426.3
OS: Linux - 5.4.0-169-generic
Java: 21.0.1 - Private Build (OpenJDK 64-Bit Server VM)
---
active-directory:2.35
ansicolor:1.0.4
antisamy-markup-formatter:162.v0e6ec0fcfcf6
apache-httpcomponents-client-4-api:4.5.14-208.v438351942757
asm-api:9.7-33.v4d23ef79fcc8
audit-trail:361.v82cde86c784e
authentication-tokens:1.53.v1c90fd9191a_b_
bootstrap5-api:5.3.3-1
bouncycastle-api:2.30.1.77-225.v26ea_c9455fd9
branch-api:2.1169.va_f810c56e895
build-failure-analyzer:2.5.1
build-monitor-plugin:1.14-876.v2162064db_a_93
build-name-setter:2.4.2
build-timeout:1.32
build-user-vars-plugin:166.v52976843b_435
caffeine-api:3.1.8-133.v17b_1ff2e0599
checks-api:2.2.0
cloudbees-disk-usage-simple:203.v3f46a_7462b_1a_
cloudbees-folder:6.858.v898218f3609d
command-launcher:107.v773860566e2e
commons-lang3-api:3.13.0-62.v7d18e55f51e2
commons-text-api:1.11.0-109.vfe16c66636eb_
copyartifact:722.v0662a_9b_e22a_c
credentials:1337.v60b_d7b_c7b_c9f
credentials-binding:677.vdc9d38cb_254d
display-url-api:2.204.vf6fddd8a_8b_e9
echarts-api:5.5.0-1
extended-read-permission:53.v6499940139e5
extra-columns:1.26
font-awesome-api:6.5.2-1
generic-webhook-trigger:2.0.0
git:5.2.2
git-client:4.7.0
gitlab-plugin:1.8.1
gson-api:2.10.1-15.v0d99f670e0a_7
hudson-wsclean-plugin:1.0.8
instance-identity:185.v303dc7c645f9
ionicons-api:74.v93d5eb_813d5f
jackson2-api:2.17.0-379.v02de8ec9f64c
jakarta-activation-api:2.1.3-1
jakarta-mail-api:2.1.3-1
javadoc:243.vb_b_503b_b_45537
javax-activation-api:1.2.0-6
javax-mail-api:1.6.2-9
jaxb:2.3.9-1
jdk-tool:73.vddf737284550
jersey2-api:2.42-147.va_28a_44603b_d5
jobConfigHistory:1229.v3039470161a_d
joda-time-api:2.12.7-29.v5a_b_e3a_82269a_
jquery3-api:3.7.1-2
jsch:0.2.16-86.v42e010d9484b_
json-api:20240303-41.v94e11e6de726
json-path-api:2.9.0-58.v62e3e85b_a_655
junit:1265.v65b_14fa_f12f0
ldap:725.v3cb_b_711b_1a_ef
leastload:3.0.0
mailer:470.vc91f60c5d8e2
matrix-auth:3.2.2
matrix-project:822.824.v14451b_c0fd42
maven-plugin:3.23
metrics:4.2.21-449.v6960d7c54c69
mina-sshd-api-common:2.12.1-101.v85b_e08b_780dd
mina-sshd-api-core:2.12.1-101.v85b_e08b_780dd
outbound-webhook:0.3.0+nmu1
pam-auth:1.10
parameterized-trigger:806.vf6fff3e28c3e
plain-credentials:182.v468b_97b_9dcb_8
plugin-usage-plugin:4.4
plugin-util-api:4.1.0
publish-over:0.22
publish-over-ssh:1.25
rebuild:332.va_1ee476d8f6d
resource-disposer:0.23
role-strategy:717.v6a_69a_fe98974
saferestart:0.7
scm-api:690.vfc8b_54395023
script-security:1336.vf33a_a_9863911
slack:715.v1cfed1b_9c63c
snakeyaml-api:2.2-111.vc6598e30cc65
ssh-credentials:337.v395d2403ccd4
ssh-slaves:2.948.vb_8050d697fec
sshd:3.322.v159e91f6a_550
structs:337.v1b_04ea_4df7c8
syslog-logger:1.0.5
timestamper:1.27
token-macro:400.v35420b_922dcb_
trilead-api:2.142.v748523a_76693
variant:60.v7290fc0eb_b_cd
versioncolumn:233.v2d198f8212a_2
workflow-api:1311.v4250456a_e552
workflow-job:1385.vb_58b_86ea_fff1
workflow-scm-step:427.v4ca_6512e7df1
workflow-step-api:657.v03b_e8115821b_
workflow-support:907.v6713a_ed8a_573
ws-cleanup:0.45
```

**Comment by nuxi** [comment]
I was able to narrow down the version range. Version 2.1.1 of the plugin works with jobs created by the REST API. Version 2.2.0 of the plugin does not see jobs created by the REST API until they are opened and re-saved.

I forgot to mention in the original report that these webhooks are using token auth. But that seems like it might be relevant given the changelog for v2.2.0

**Comment by tomasbjerre** [comment]
So 2.2.1 works without any problems? And that means this is not an issue in the latest release?

Sounds strange that i works in 2.2.1 but not in 2.2.0, if you look at what changed:
https://github.com/jenkinsci/generic-webhook-trigger-plugin/compare/generic-webhook-trigger-2.2.0...generic-webhook-trigger-2.2.1

If it actually works in 2.2.1 and not in 2.2.0, I think you have some other difference in your setup (some other plugin version might differ).

**Comment by nuxi** [comment]
> So 2.2.1 works without any problems? And that means this is not an issue in the latest release?

No, v2.2.1 and v2.2.0 are both broken.

Older versions like v2.0.0 and v2.1.1 both work fine.

**Comment by tomasbjerre** [comment]
If itemListener is not invoked when creating jobs with the Jenkins rest API, I think this is a big in Jenkins core.

This plugin should be invoked here:
https://github.com/jenkinsci/generic-webhook-trigger-plugin/blob/3954ee99f8c8801877f6a9b998fe7fdcaa78e6d7/src/main/java/org/jenkinsci/plugins/gwt/jobfinder/JobFinderImpersonater.java#L20

**Comment by nuxi** [comment]
> If itemListener is not invoked when creating jobs with the Jenkins rest API, I think this is a big in Jenkins core.

Is there any chance you could add a workaround or change the code back (this worked in previous versions of the plugin) until Jenkins core is fixed?

**Comment by tomasbjerre** [comment]
> > If itemListener is not invoked when creating jobs with the Jenkins rest API, I think this is a big in Jenkins core.
> 
> Is there any chance you could add a workaround or change the code back (this worked in previous versions of the plugin) until Jenkins core is fixed?

Changing it back would just cause other problems. I asked for advice here:
https://groups.google.com/g/jenkinsci-dev/c/w-jdCAtGh20/m/k6Gvl7P9AwAJ

**Comment by mawinter69** [comment]
I think this is a bug in core. When posting an xml via the rest api or the cli to update a job, the ItemListener is not called, only a SaveableListener. And the bug for #311 is basically the same as JobDSL internally is calling the same method to update the xml.

**Comment by mawinter69** [comment]
I opened two PRs in core that will fix the mentioned problems:
https://github.com/jenkinsci/jenkins/pull/9304 to fix the issue from JENKINS-64553 which also solves this issue
https://github.com/jenkinsci/jenkins/pull/9303 to fix the issue from JENKINS-64356

**Comment by mawinter69** [comment]
Fix for ItemListener will be included in next LTS 2.452.2

**Comment by tomasbjerre** [comment]
Continuing this in #319
